### ISSUE TITLE
⚙Tech Debt: Distinguish between elements, stores, & actions

### ISSUE THREAD

**Issue by huntabyte** [melt-ui/melt-ui#231]
As the library stands right now, to create a dialog, you would the following:

```ts
const { trigger, portal, overlay, content, title, description, close, open } = createDialog();
```

The problem is that it's not easy to distinguish between what should be applied to an element, what is just a state store, what is an action, etc.

The issue becomes more obvious for more complex components that may have elements, actions, state stores, call expressions with dependencies, etc.

We've discussed this internally but wanted to open it up to the community for input/feedback on this as well.

The initial proposal is to add an extra layer to the return object, which might look something like this:

```ts
const { elements, actions, states } = createDialog()
```

You could destructure each category individually, like so:
```ts
const { elements, actions, states } = createDialog()
const { trigger, overlay, content, title, description, close } = elements
const { portal } = actions
const { open } = states
```
or do it all in one go if you really wanted

```ts
const { elements: { trigger, overlay, content, title, description, close }, actions: { portal }, states: { open } } = createDialog()
```

While this does add some work to the dev/user, it's also a form of self-documentation, as it becomes more obvious what the purpose of each of these is. Especially in this example if you consider the `close` and `open`. At first glance, `close` seems like it should be a function that when called, closes the menu, but it's an element. By adding that extra layer we remove that confusion altogether. "Okay, `close` is a part of elements so it must be applied to an element, `open` is a state, so it must be tracking the open state of the dialog, `portal` is an action, so I need to prefix it with `use:`".

For something like the `Select` builder, which has an `isSelected` that is returned, we could categorize those types of expressions under `helpers` (I'm the worst at naming things so feedback would be nice here).

It could go from:
```ts
const { label, trigger, menu, open, option, isSelected, group, groupLabel } = createSelect();
```

to:
```ts
const { elements, helpers, states } = createSelect();
const { label, trigger, menu, option, group, groupLabel } = elements
const { isSelected } = helpers
const { open, options } = states
```

**Comment by saturnonearth** [comment]
Yea definitely separate things like stores from elements. Since Svelte doesn't provide a clear way to tell otherwise. That or adopt a naming convention like `sOpen` or `openStore` to better differentiate them.

**Comment by isaacfink** [comment]
Personally I don't see it as an issue (but I've worked extensively with melt so it doesn't matter what I say) but maybe naming things might be a good solution for example `triggerEl` and `valueStore` ots more verbose but still one level for destructuring

**Comment by huntabyte** [comment]
@isaacfink the problem is that `valueStore` and `triggerEl` are both _technically_ stores 🤣

**Comment by isaacfink** [comment]
@huntabyte that's true, I didn't mean this convention specifically just the idea of adding a postfix to differentiate this way it doesn't change the structure (and fwiw its also backwards compatible)

**Comment by isaacfink** [comment]
And isn't everything a store? We might have to come up with a name for values, or maybe even stick with store and the meaning is anything that is not a helper or element

**Comment by huntabyte** [comment]
yeah, basically everything is a store @isaacfink, which is I why I used `state`, rather than `stores` for the property name!

**Comment by Klustre** [comment]
I'm looking to pick up Melt (actually Shadcn Svelte, but not until it has integrated Melt) and I haven't tried either yet. But I don't see the need to group imports. The docs and types should detail the specifics and the examples for each component should be explanatory enough. I wouldn't try to solve this issue until the library has greater adoption.

**Comment by TGlide** [comment]
While the post-fix is clear, it can be quite verbose. e.g. `checkboxItemElement`. It also doesn't have the organization that this does:

```ts
const { 
  elements: { label, trigger, menu, option, group, groupLabel }, 
  helpers: { isSelected },
  states: { open, options } 
} = createSelect();
```

But after you get out of the script scope and into the template, the clarity is gone. What @huntabyte suggested is we use some JSDoc to add some info to when you're hovering the variable, which I think is an excellent idea.

**Comment by TGlide** [comment]
> I'm looking to pick up Melt (actually Shadcn Svelte, but not until it has integrated Melt) and I haven't tried either yet. But I don't see the need to group imports. The docs and types should detail the specifics and the examples for each component should be explanatory enough. I wouldn't try to solve this issue until the library has greater adoption.

I'd rather solve this before it has greater adoption, really! The design for Melt's API should not change between major versions. We can add stuff and deprecate them on other versions, but I'd rather deal with this sooner so things aren't messy later on.

**Comment by geerew** [comment]
I have absolutely had the trouble of knowing what has an action, what is setting props what is a plain old store (like $open). 

I like the suggested approach by @TGlide and @huntabyte. While it’s (potentially) more likes of code, it makes it absolutely clear what is what and that’s the most important thing when developing or when looking back on code in a years time

I also love the JSDOC idea. Too many times I hover on a restructured value and it gives me no indication of what it is and I’m right back to looking at the docs

**Comment by huntabyte** [comment]
@geerew thanks for the feedback! I know @AdrianGonz97 is a JSDoc wizard so going to seek his guidance on ensuring the JSDocs are *chef's kiss*.

**Comment by huntabyte** [comment]
> > I'm looking to pick up Melt (actually Shadcn Svelte, but not until it has integrated Melt) and I haven't tried either yet. But I don't see the need to group imports. The docs and types should detail the specifics and the examples for each component should be explanatory enough. I wouldn't try to solve this issue until the library has greater adoption.
> 
> I'd rather solve this before it has greater adoption, really! The design for Melt's API should not change between major versions. We can add stuff and deprecate them on other versions, but I'd rather deal with this sooner so things aren't messy later on.

💯 - we've already begun to have issues/support questions raised that could have potentially been avoided had it been more clear. The feedback requested is more along the lines of the names of things/what makes the most sense to the most people!

**Comment by Klustre** [comment]
> I'd rather solve this before it has greater adoption, really! The design for Melt's API should not change between major versions. We can add stuff and deprecate them on other versions, but I'd rather deal with this sooner so things aren't messy later on.

That goes without saying. But until it hits 1.0 you can wait to see if users regularly get confused with the imports.

**Comment by TGlide** [comment]
> > I'd rather solve this before it has greater adoption, really! The design for Melt's API should not change between major versions. We can add stuff and deprecate them on other versions, but I'd rather deal with this sooner so things aren't messy later on.
> 
> That goes without saying. But until it hits 1.0 you can wait to see if users regularly get confused with the imports.

We've seen some confusion arising already, and sometimes even I get confused by it if I'm being honest.

**Comment by BenocxX** [comment]
I like this idea! 

Instead of `helpers`, how about `context`? I don't know any other helpers from the top of my mind, but isSelected is more context related than an actual helper. I feel like it would make more sense and could be reused for other properties like `open` from `createDialog`.

**Comment by TGlide** [comment]
> I like this idea!
> 
> Instead of `helpers`, how about `context`? I don't know any other helpers from the top of my mind, but isSelected is more context related than an actual helper. I feel like it would make more sense and could be reused for other properties like `open` from `createDialog`.

I'd say context would only make sense if we replaced `state` and `helpers` with it 🤔 Especially since `open` is a state currently.

The separation currently is interesting, because state is associated directly with stores we use internally in the builders. While helpers we normally don't use internally, but expose them to help consume the builders
